### PR TITLE
[Bug]Fix label smooth bug

### PR DESCRIPTION
**Pull request by HIT-cwh**
Hi,
This pr fix a bug in cross_entropy_loss.py/soft_cross_entropy, and implement a new LabelSmoothLoss.
Pls kindly take a look.
Thank you.

### CODE REVIEW

File: mmcls/models/losses/label_smooth_loss.py [hunk 1]
```diff
@@ -1,50 +1,65 @@
...
-    return loss
+from .cross_entropy_loss import CrossEntropyLoss
+from .utils import convert_to_one_hot
 
 
 @LOSSES.register_module()
-class LabelSmoothLoss(nn.Module):
+class LabelSmoothLoss(CrossEntropyLoss):
+    """Intializer for the label smoothed cross entropy loss.
+        This decreases gap between output scores and encourages generalization.
```
Comment (MeowZheng) [inline mmcls/models/losses/label_smooth_loss.py:11]: blank line before 'This decreases gap...

File: mmcls/models/losses/label_smooth_loss.py [hunk 2]
```diff
@@ -1,50 +1,65 @@
...
+        Labels provided to forward can be one-hot like vectors (NxC) or class
+        indices (Nx1).
+        This normalizes the labels to a sum of 1 based on the total count of
+        positive targets for a given sample before applying label smoothing.
+
+        Args:
+            label_smooth_val: value to be added to each target entry
+            reduction: specifies reduction to apply to the output
+        """
```
Comment (MeowZheng) [inline mmcls/models/losses/label_smooth_loss.py:20]: there is no indentation. 

Args must contain the type, function, default values... of all arguments in __init__ function. 

please refer to https://github.com/open-mmlab/mmclassification/blob/master/mmcls/models/losses/focal_loss.py#L55

File: mmcls/models/losses/label_smooth_loss.py [hunk 3]
```diff
@@ -1,50 +1,65 @@
...
+    def generate_one_hot_like_label(self, label):
+        """
+        This function takes one-hot or index label vectors and computes
+        one-hot like label vectors (float)
+        """
+        label_shape_list = list(label.size())
+        # check if targets are inputted as class integers
+        if len(label_shape_list) == 1 or (len(label_shape_list) == 2
+                                          and label_shape_list[1] == 1):
+            label = convert_to_one_hot(label.view(-1, 1), self.num_classes)
```
Comment (MeowZheng) [inline mmcls/models/losses/label_smooth_loss.py:45]: it might be better to throw an exception when label shape is not (C,) or (C,1)

File: mmcls/models/losses/label_smooth_loss.py [hunk 4]
```diff
@@ -53,16 +68,24 @@ def forward(self,
...
+                f'num_classes should equal to cls_score.shape[1], ' \
+                f'but got num_classes: {self.num_classes} and ' \
+                f'cls_score.shape[1]: {cls_score.shape[1]}'
+        else:
+            self.num_classes = cls_score.shape[1]
+        one_hot_like_label = self.generate_one_hot_like_label(label=label)
+        assert (
+            one_hot_like_label.shape == cls_score.shape
+        ), 'LabelSmoothingCrossEntropyLoss requires output and ' \
+           'target to be same size'
```
Comment (MeowZheng) [inline mmcls/models/losses/label_smooth_loss.py:82]: it might be better to add output shape and target shape in error msg

File: mmcls/models/losses/utils.py [hunk 5]
```diff
@@ -96,3 +97,17 @@ def wrapper(pred,
         return loss
 
     return wrapper
+
+
+def convert_to_one_hot(targets: torch.Tensor, classes) -> torch.Tensor:
+    """
+    This function converts target class indices to one-hot vectors,
+    given the number of classes.
+    """
```
Comment (MeowZheng) [inline mmcls/models/losses/utils.py:106]: add Args

File: mmcls/models/losses/label_smooth_loss.py [hunk 6]
```diff
@@ -1,50 +1,69 @@
...
+    indices (Nx1).
+    This normalizes the labels to a sum of 1 based on the total count of
+    positive targets for a given sample before applying label smoothing.
 
-
-@LOSSES.register_module()
-class LabelSmoothLoss(nn.Module):
+    Args:
+        label_smooth_val (float): value to be added to each target entry
+        num_classes (int): the number of classes. Defaults to None.
```
Comment (MeowZheng) [inline mmcls/models/losses/label_smooth_loss.py:20]: as usual, if the defaulted value is None, the argument is optional.  
like....num_classes (int, optional): the number of classes. Defaults to None.

**Review by MeowZheng — APPROVED** [review]
